### DISCUSSION TITLE
OnTap Popup on map

### DISCUSSION THREAD

**Discussion by gregws2** [mapsforge/mapsforge#1229] [Q&A] [answered]
Hiyas, I've been working on getting a popup with details on tap with FixedPixelCircle but I am having issues with it showing when the points are hidden.

The map correctly makes layers disappear using
`mapView.getLayerManager().getLayers().get(row).setVisible(toggle,true);`
but when testing in the onTap for the FixedPixelCircle's the Polygon still appears.

```
@Override
        public boolean onTap(LatLong tapLatLong, Point layerXY, Point tapXY) {
                Point circleXY = GeneralMapping.mapView.getMapViewProjection().toPixels(getPosition());

                if(polygon != null) { 
                    GeneralMapping.mapView.getLayerManager().getLayers().remove(polygon);
                    polygon = null;
                }
                
                if(this.contains(circleXY, tapXY) && this.isVisible()) {

                    Paint paint = GeneralMapping.GF.createPaint();
                    paint.setColor(AwtGraphicFactory.INSTANCE.createColor(200, 10, 10, 10));

                    Paint paintFill = GeneralMapping.GF.createPaint();
                    paintFill.setColor(AwtGraphicFactory.INSTANCE.createColor(40, 10, 10, 10));

                    paint.setStyle(Style.STROKE);
                    paint.setStrokeWidth(2); // Set the strike width here

                    polygon = new Polygon(paintFill, paint, GeneralMapping.GF);

                    List<LatLong> latLongs3 = new ArrayList<>();
                    System.out.println(tapLatLong.getLatitude());
                    System.out.println(tapLatLong.getLongitude());
                    latLongs3.add(new LatLong(tapLatLong.getLatitude(), tapLatLong.getLongitude()));
                    latLongs3.add(new LatLong(tapLatLong.getLatitude()-(0.0003*32768/10000), tapLatLong.getLongitude()));
                    latLongs3.add(new LatLong(tapLatLong.getLatitude()-(0.0003*32768/10000), tapLatLong.getLongitude()+0.0008*32768/10000));
                    latLongs3.add(new LatLong(tapLatLong.getLatitude(), tapLatLong.getLongitude()+0.0008*32768/10000));
                    latLongs3.add(new LatLong(tapLatLong.getLatitude(), tapLatLong.getLongitude()));
                    
                    polygon.setPoints(latLongs3);
                    
                    GeneralMapping.mapView.getLayerManager().getLayers().add(polygon);

                    return true;
                }
                
            return false;
        }
```
Click on the visible point and the box appears
![image](https://user-images.githubusercontent.com/57698218/125939036-06b59cf7-081e-4727-88e4-0456c7ec4613.png)

Click on a point that has been rehidden and a box still appears - Note that there is a point on the map, just the layer has been hidden
![image](https://user-images.githubusercontent.com/57698218/125939146-058e7f6d-4f91-4612-a7c0-6cacbcd20e81.png)

My intention would be for the Polygon Box to only appear when the point is visible.

Should I be using something different to Polygons? I'm using AWT in desktop application. Are bubbles only for Android? I will eventually want to put some parameters in the box when clicked.

Thanks for any help you can provide.

**Comment by devemux86** [comment]
> My intention would be for the Polygon Box to only appear when the point is visible.

It is not clear what you mean.
You show / hide the layers, so you can adjust their visibility based on your actions.

Or check the `isVisible` response in `onTap` method and return `false` to skip the event.

> Are bubbles only for Android?

You can create bitmaps from Java panels and set them as markers.
(similar to the process with [Android text views](https://github.com/mapsforge/mapsforge/blob/master/mapsforge-samples-android/src/main/java/org/mapsforge/samples/android/BubbleOverlay.java))

**Reply by gregws2** [reply]
Thanks for the response.
So in the first image, you can see the point, when clicked it shows the polygon. This is expected.
In the second image, the layer isVisible() = false, but the polygon still shows when I click on "empty space".

Do I need to set each FixedPixelCircle's visibility on the layer that I hide, or do I just need to hide the layer?

I would have expected the `if(this.contains(circleXY, tapXY) && this.isVisible()) {` to not trigger if the layer is hidden.

*edit*
When checking FixedPixelCircle's this.isVisible() within the onTap it is true for points after using `mapView.getLayerManager().getLayers().get(row).setVisible(false,true);`

Although within the draw of the FixedPixelCircle, this.isVisible() is acting correctly.

**Reply by devemux86** [reply]
You can add / remove a layer when changing its visibility.
Or you need to check its visibility inside its event methods.

You can override the `contains` method and adjust it based on your app needs.
Each app has different requirements, so contains / visible can be used differently.

**Comment by devemux86** [comment, marked as answer]
> I would have expected the if(this.contains(circleXY, tapXY) && this.isVisible()) { to not trigger if the layer is hidden.

The `contains` just checks geometries, so it is valid.

The visibility check belongs in the overridden events of each layer.
(according to each app requirements)

**Reply by gregws2** [reply]
Thanks for helping Emux.
I managed to get it to work.

I added a `private boolean onMap;` to each circle and a reference point back to my address points array.
When I set the mapview layer visibility to false, I also alter the visibility of points in the address point array to false.

Then `onTap` I checked the visibility state of my points array to see the visibility state that the circle should be in.
`this.onMap = listOfAddressPoints.get(this.referencePoint).getVisibleState();`

Probably not the ideal method but it works.

Thanks again
